### ISSUE TITLE
(Non-Input) Text Selection 

### ISSUE THREAD

**Issue by dvdwasibi** [flutter/flutter#5422]
It would be useful to be able to create Text widgets or RichText widgets that are selectable by users.

Maybe a "selectable" flag in the RichText widget would suffice.

**Comment by shashikantdurge** [comment]
Any updates on this ?
When can we expect it to be live?

**Comment by wendux** [comment]
Any updates on this ?
When can we expect it to be live?

**Comment by CalebQ42** [comment]
For now, I've implemented a solution that isn't too difficult. Basically I just downloaded the files for EditableText and TextField, then stripped out the code that makes a keyboard show up. Not as fancy as a richtext, but works for my needs.

**Comment by marianoarga** [comment]
@CalebQ42 404 on both URL

**Comment by CalebQ42** [comment]
@marianoarga Oops, had moved them for organization. I updated the links.

**Comment by jifalops** [comment]
@CalebQ42 can you describe what your widget can/cant do as far as text selection goes?

**Comment by CalebQ42** [comment]
@jifalops They work exactly like a TextField or EditableText, just that when they become focused they don't popup the keyboard. I haven't even gotten around to removing unnecessary variable so all the constructor variables and methods are still there. It's a quick and dirty solution, but works for what I need.

**Comment by lifenautjoe** [comment]
It's been a couple of months. Has anyone come up with a solution to this?

**Comment by ayush221b** [comment]
Hi, is there a solution available for this. I really need to be able to select a specific word, and my issue was closed as a duplicate of this issue. Is there a workaround, or a hack, or something that can be done?

**Comment by Karnor** [comment]
Not that I am aware off. It's one of the main reasons I have stopped using flutter, and gone back to android java coding.The devs are 100% wrong when they class this as nice to have but not needed.

    On Tuesday, 15 January 2019, 10:13:48 GMT, Ayush Shekhar <notifications@github.com> wrote:  
 
 
Hi, is there a solution available for this. I really need to be able to select a specific word, and my issue was closed as a duplicate of this issue. Is there a workaround, or a hack, or something that can be done?

**Comment by zoechi** [comment]
>  when they class this as nice to have but not needed.

Nobody did that though

**Comment by Karnor** [comment]
I seem to remember seeing that in the bug listing, could be wrong, might have been elsewhere, but not in a position to check atm.

    On Tuesday, 15 January 2019, 14:12:50 GMT, Günter Zöchbauer <notifications@github.com> wrote:  
 
 

when they class this as nice to have but not needed.


Nobody did that though

**Comment by Karnor** [comment]
hmm, I Might have misunderstood that it was the requestor marking it as such, in response to the devs question of severity
Ability to select and copy part of text · Issue #6887 · flutter/flutter

| 
| 
| 
|  |  |

 |

 |
| 
|  | 
Ability to select and copy part of text · Issue #6887 · flutter/flutter

When using Text or RichText widgets, resulting text could not be selected and copied. It would be nice to have a...
 |

 |

 |



But I seem to remember there being another thread with this in where I took part in the discussion, so if I can ever work out how to retrieve my posts on github, I'll have a look for it to check I did not make that same mistake twice. None of the comment queries I make show the thread where I commented on this. It's not the same thread as this one.

    On Tuesday, 15 January 2019, 14:12:50 GMT, Günter Zöchbauer <notifications@github.com> wrote:  
 
 

when they class this as nice to have but not needed.


Nobody did that though

**Comment by bgetsug** [comment]
This seems to work for me:

```dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

final someText = "abc123";

TextFormField selectOnly() {
  return TextFormField(
    decoration: InputDecoration(labelText: "ID"),
    initialValue: someText,
    inputFormatters: [
      TextInputFormatter.withFunction((oldValue, newValue) {
        return TextEditingValue(
          text: oldValue.text,
          selection: newValue.selection,
        );
      })
    ],
  );
}
```

**Comment by 00SunnyDay00** [comment]
We do have received many complaints from our users.  Does the flutter team have any plan to  support this feature?

**Comment by sdresche17** [comment]
Desperately waiting for this feature, can the devs at least provide an update?

**Comment by msal4** [comment]
Is there anyone working on it?

**Comment by lifenautjoe** [comment]
@msal4 It's place on "Future" milestones.

> P4: These are for features we are interested in but explicitly punting for the foreseeable future. Typically this is for features that go beyond the scope of our initial goal of having a solid UI framework for making applications for Android and iOS phones and tablets. We are extremely unlikely to work on issues in this milestone. 

Meaning they don't think this is part of having a solid UI framework for making apps and won't be worked on anytime soon. 🤦‍♂️

**Comment by miguelcaravantes** [comment]
I need this too

**Comment by Hixie** [comment]
I would expect this to be implemented by refactoring EditableText such that there is a simpler version of it that just implements selection (and EditableText would be based on it).

...or maybe just a way to make EditableText readonly, if the refactoring seems intractable.

**Comment by docaohuynh** [comment]
Is Flutter 1.2.1 include this issue?

**Comment by zoechi** [comment]
@docaohuynh the issue is still open (see top-left) and therefore a fix not included anywhere

**Comment by msal4** [comment]
@lifenautjoe too bad! I think I'm gonna work on a package that implements this behaviour for the mean time.

**Comment by ayush221b** [comment]
> @lifenautjoe too bad! I think I'm gonna work on a package that implements this behaviour for the mean time.

Thank you so much sir. Would be very grateful.

**Comment by rishabh-chimple** [comment]
@msal4 yes Please.Thanks that will be very much grateful.

**Comment by satish754ss** [comment]
Thank you so much. @msal4 .

**Comment by stephensamonte** [comment]
You can save text to the clipboard with the service package. 

This is a possible solution: https://stackoverflow.com/questions/46260055/how-to-make-copyable-text-flutter-widget

This copies an entire block of text:
```
import 'package:flutter/services.dart';

String textToCopy = "test";
    return new GestureDetector(
      child: new Text(textToCopy),
      onLongPress: () {
        Clipboard.setData(new ClipboardData(text: textToCopy));
        key.currentState.showSnackBar(
            new SnackBar(content: new Text("Copied to Clipboard"),));
      },
    );
```

**Comment by lifenautjoe** [comment]
Hi @msal4 ! How did that go? Were you able to get somewhere? 😬

**Comment by Smagold** [comment]
any updates?

**Comment by Smagold** [comment]
Tried using `RenderEditable` with a controller based on `TextEditingValue`. That seems to be the only object that handles text selection in Flutter.

Selection appears on long tap, though I'm unable to change it by dragging the handles. It seems that without an active `TextInput` (active keyboard), selection object does not seem to get updated.

I was able to get "Select All" and one-word selection to work, as well as "Copy" and "Paste" & "Cut" (could not remove these 2, I only needed copy).

Still, so far the whole thing is barely usable. I doubt that `RenderEditable` is a way to go.

How come such simple UX functionality as text selection is not supported in Flutter?!

**Comment by dannyvalentesonos** [comment]
I've actually got a SelectableText widget that uses a EditableText like the TextField does, but also uses a subclass of a FocusNode so that it doesn't consume the keyboard token and it makes the EditableText widget do most of the things needed, but it still shows a Cut and Paste action in the toolbar, which isn't ideal.
Also, this doesn't allow for multiple text spans unfortunately...
If anyone would like to see the solution I have, I could share here.

**Comment by ayush221b** [comment]
@dannyvalentesonos Hi, could you please share the code.

**Comment by dannyvalente** [comment]
@ayush221b I have published a flutter package.
https://pub.dartlang.org/packages/flutter_selectable_text/

Let me know if you have any questions.

**Comment by dander11** [comment]
@dannyvalente Thanks for that, great job! 

However, your package doesn't support any RichText elements. Which I'm sure is a lot of use cases. So while your package is definitely a great first step, this is definitely still a big issue with Flutter. 

Do you have any plan to support RichText? I imagine that would be a pretty big task.

**Comment by dnfield** [comment]
We want to have support for text selection handles in what are effectively read-only text fields.

I suspec tthere's a class of use cases where this is just a matter of wrapping a `Text` widget with a `GestureDetector` that provides a context menu or dialog for copying (or, just flat out copies the text).  There are many apps that do this today (e.g. Twitter app for Android, Messages).

**Comment by dannyvalente** [comment]
> @dannyvalente Thanks for that, great job!
> 
> However, your package doesn't support any RichText elements. Which I'm sure is a lot of use cases. So while your package is definitely a great first step, this is definitely still a big issue with Flutter.
> 
> Do you have any plan to support RichText? I imagine that would be a pretty big task.

My package intentionally only supports regular Text elements as stated in the description. Unfortunately, it would be a large undertaking to try to support that, I'd hope the Flutter team will add that eventually.

**Comment by dannyvalente** [comment]
> We want to have support for text selection handles in what are effectively read-only text fields.
> 
> I suspect there's a class of use cases where this is just a matter of wrapping a `Text` widget with a `GestureDetector` that provides a context menu or dialog for copying (or, just flat out copies the text). There are many apps that do this today (e.g. Twitter app for Android, Messages).

Unfortunately, I don't think this kind of solution would be enough though. How would one write an email app for example using Flutter, without actually having the ability to let the user select text, including different styles throughout said text?

**Comment by dnfield** [comment]
@dannyvalente I agree that we need the full fledged selection of read-only text support with selection handles and support for styles.   I'm just suggesting that some people looking for this would be well served with a simpler/lighter weight solution.

**Comment by dannyvalente** [comment]
@dnfield Thanks for clarifying. Your previous post made it sound like the idea was to do something similar to what I did with text fields, and I was wondering if that meant only support for standard text.
Thanks again!

**Comment by guilhermedecampo** [comment]
Have something like medium or kindle app would be amazing!

![IMG_5490](https://user-images.githubusercontent.com/4806269/56740132-b5624a00-6746-11e9-951c-9ab63d4f9a42.PNG)
![IMG_5491](https://user-images.githubusercontent.com/4806269/56740133-b5fae080-6746-11e9-8093-9acfa2b9cbb0.PNG)

**Comment by msal4** [comment]
Sorry @lifenautjoe, I had an unfortunate accident a while ago and I intend on contributing to what others have worked on, ass soon as I recover.

**Comment by KNightING** [comment]
> @ayush221b I have published a flutter package.
> https://pub.dartlang.org/packages/flutter_selectable_text/
> 
> Let me know if you have any questions.

Thank your share and I refer to this to make my [selectable text](https://pub.dartlang.org/packages/flutter_selectext).

I published and working but I don't know how to solve health issues and detected platforms....

**Comment by mattarau** [comment]
I think this should now have its priority status updated since on the web (with `flutter_web`) selectable texts are not a "nice to have", but a requirement.

**Comment by chunhtai** [comment]
We have been actively working on this, currently it is scheduled to be done by end of May unless further notice.

**Comment by chunhtai** [comment]
This issue is pushed to end of June. There are multiple steps that need to happen before this can be implemented, the https://github.com/flutter/flutter/issues/14014 is the first step which is what I have been working on. Once that fixed, we can use it as a temporary work around for selectable text. What is left for this issue will be writing a new wrapper widget with api more close to text widget.

**Comment by dander11** [comment]
Thanks for the update @chunhtai . It makes sense that this would take a while since there are so many types of text widgets that would need to be supported! Good luck with it.

**Comment by Mayaela** [comment]
@chunhtai any news ?

**Comment by chunhtai** [comment]
#14014 is merged. For now you can micmic the selectable text by
```
TextField(
    controller: TextEditingController(text: 'initial value),
    readOnly: true,
    maxLines: null,
    decoration: null
)
```
it currently does not support rich text. and it always take up full width. I will continue work on the selectable text with easier to use api and more functionality

**Comment by PieterAelse** [comment]
TIL that text **IS selectable** in the Android app-switcher!?

Would Flutter devs please like to explain why this happens why any normal Text widget "doesn't support" text selection, but in Androids app-switcher it does?

[Demo movie: CLICK](https://www.dropbox.com/s/yd9sj4r93yuc5tw/Flutter_TextSelection.mp4?dl=0)

**Comment by creativecreatorormaybenot** [comment]
@PieterAelse I bet that Android has some text recognition software that allows this.

**Comment by dannyvalentesonos** [comment]
That is a feature in Android N or O that uses text recognition to allow selection there.

**Comment by guilhermedecampo** [comment]
@chunhtai 	how can you explain very quickly how to make a Text selectable after this merge?

**Comment by chunhtai** [comment]
It seems the api doc is not propagated yet,
it is in package:flutter/material.dart
to use it 
```
SelectableText('magic')
```
The api is pretty similar to text

**Comment by creativecreatorormaybenot** [comment]
@chunhtai I cannot find it on master. Is there a pull request for this?

**Comment by dander11** [comment]
Just a side note for anyone else not seeing SelectableText in their material package. Make sure you're on the master branch, this change is not on dev or beta yet, and run flutter upgrade to get the change. 

Thanks for all your work @chunhtai !!

**Comment by ADLsourceCode** [comment]
@dander11  @chunhtai  I am using the below flutter version:-
Flutter 1.7.8+hotfix.4 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 20e59316b8 (2 weeks ago) • 2019-07-18 20:04:33 -0700
Engine • revision fee001c93f
Tools • Dart 2.4.0


I can not see the SelectableText widget here... Pls Help

**Comment by KagurazakaHanabi** [comment]
@ADLsourceCode 
```
flutter channel master
flutter upgrade
```

**Comment by ADLsourceCode** [comment]
@KagurazakaHanabi  @dander11  @chunhtai  thanks .. I found it.. Is there a way to overide the default menu of **copy** and **selectALL**...

**Comment by KagurazakaHanabi** [comment]
@ADLsourceCode No...See #31412

**Comment by ADLsourceCode** [comment]
> 
> 
> @ADLsourceCode No...See #31412

@KagurazakaHanabi any chance of getting it added.... any alternative method you can recommend to make the app work... as this is important for my app

**Comment by chunhtai** [comment]
@ADLsourceCode please use thumb up if you want to bump up the priority

**Comment by ADLsourceCode** [comment]
@chunhtai  by when can we expect this to happen. any idea...

**Comment by ejamesc** [comment]
@chunhtai thank you so much for your work on this. I’m very grateful. 

@ADLsourceCode please thumbs up on #31412, and move all discussion there.

**Comment by sanekyy** [comment]
when it will supported in flutter web?

**Comment by sanekyy** [comment]
it's very necessary feature for web

**Comment by yjbanov** [comment]
We do want this feature on the web. It is harder to say when we'll be able to start actual work.

**Comment by sanekyy** [comment]
Okay, thanks. I'll wait
